Kuadrant/limitador-operator
# pull an image from a private registry

### What

Pull an image from a private registry

Fixes #78 

### Verification steps

First, get some limitador image on a private repository. For example: `quay.io/eastizle/limitador-private:v1.6.0` (this repo and image will be removed when this PR is merged).

dev setup

```
make local-setup
```

Create a Secret of type `kubernetes.io/dockerconfigjson by providing credentials. For example, using kubectl tool with the following command line:

```
kubectl create secret docker-registry regcred --docker-server=<your-registry-server> --docker-username=<your-name> --docker-password=<your-pword>
```

Deploy limitador with a custom image and a reference to the recently created `regcred` secret.

```yaml
kubectl apply -f - <<EOF
---
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-instance-1
spec:
  image: quay.io/eastizle/limitador-private:v1.6.0
  imagePullSecrets:
  - name: regcred
EOF
```
Wait for limitador to be ready

```
kubectl wait --timeout=300s --for=condition=Ready limitador limitador-instance-1
```

Check status
```
kubectl get limitador  limitador-instance-1 -o jsonpath='{.status}' | yq e -P
```

It should say `Ready`.

Now, let's deploy limitador with custom image from a private repository without providing imagepullsecrets. It will be a custom image different from the one used in the previous example. As the image of the previous example is already present in the cluster and the limitador deployment pullpolicy is "IfNotPresent", using that image even without imagepullsecrets would not fail. 

```yaml
kubectl apply -f - <<EOF
---
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador-instance-without-imagepullsecrets
spec:
  image: quay.io/eastizle/limitador-private:other
EOF
```

Check status
```
kubectl get limitador  limitador-instance-without-imagepullsecrets -o jsonpath='{.status}' | yq e -P
```
The answer should be: not "ready" with somewhat confusing message: "Deployment does not have minimum availability"

Checking pod status or events we can get more detailed information:
```
kubectl get events --field-selector reason=Failed
```

You should see `https://quay.io/v2/eastizle/limitador-private/manifests/other: 401 UNAUTHORIZED`

```
LAST SEEN   TYPE      REASON   OBJECT                                                                MESSAGE
2m11s       Warning   Failed   pod/limitador-limitador-instance-without-imagepullsecrets-6799fszk2   Failed to pull image "quay.io/eastizle/limitador-private:other": failed to pull and unpack image "quay.io/eastizle/limitador-private:other": failed to resolve reference "quay.io/eastizle/limitador-private:other": unexpected status from HEAD request to https://quay.io/v2/eastizle/limitador-private/manifests/other: 401 UNAUTHORIZED
2m11s       Warning   Failed   pod/limitador-limitador-instance-without-imagepullsecrets-6799fszk2   Error: ErrImagePull
109s        Warning   Failed   pod/limitador-limitador-instance-without-imagepullsecrets-6799fszk2   Error: ImagePullBackOff
```

### Cleanup

```
make local-cleanup
```
